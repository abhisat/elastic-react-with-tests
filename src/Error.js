import React from 'react';
import PropTypes from 'prop-types';

function Error(props) {
  const { display, notice } = props;
  const style = {
    display,
  };
  return (
    <p id="error" style={style}>{notice}</p>
  );
}

Error.propTypes = {
  display: PropTypes.string,
  notice: PropTypes.string,
};

Error.defaultProps = {
  display: 'none',
  notice: 'Error',
};
export default Error;
