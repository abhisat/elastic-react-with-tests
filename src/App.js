import React, { Component } from 'react';
import './App.css';
import ReactLoading from 'react-loading';
import Graph from './Graph';
import Error from './Error';

class App extends Component {
  constructor() {
    super();
    this.state = {
      urls: [],
      before: '',
      after: '',
      interval: '',
      submitted: false,
      loading: false,
      validInputs: false,
      labels: [],
      values: [],
      urlErrors: false,
      beforeErrors: false,
      afterErrors: false,
      intervalErrors: false,
      formErrors: false,
    };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.calculateVals = this.calculateVals.bind(this);
    this.validateInput = this.validateInput.bind(this);
  }

  // Function to take the response from API and calculate the values for the Graph
  calculateVals(response) {
    const labels = [];
    const values = [];
    let norm = 0;
    response.map((responseVal) => {
      const temp = [];
      try {
        responseVal.aggregations.by_page_views.buckets.map((bucket) => {
          if (labels.indexOf(bucket.key) < 0) {
            labels.push(bucket.key);
          }
          temp.push(bucket.doc_count);
          return 0;
        });
        values.push(temp);
        return 0;
      } catch (err) {
        temp.push(0);
        values.push(temp);
      }
      return 0;
    });
    values.map((arr) => {
      if (arr.length > 1) {
        norm = arr.length;
      }
      return 0;
    });
    values.map((arr) => {
      if (arr.length === 1) {
        for (let k = 0; k < norm - 1; k += 1) {
          arr.push(0);
        }
      }
      return 0;
    });
    this.setState({ labels });
    this.setState({ values });
    this.setState({ loading: false });
  }
  // Function to take the response from API and calculate the values for the Graph

  // Function to validate timestamp and interval inputs in the form
  validateInput(name, value) {
    if (value != null) {
      switch (name) {
        case 'urls':
          this.state.urls = value.split(',');
          this.setState({ urlErrors: false });
          break;
        case 'before':
          if (value.match(/^[0-9]+$/) || value === '') {
            this.setState({ beforeErrors: false });
          } else {
            this.setState({ beforeErrors: true });
          }
          break;
        case 'after':
          if (value.match(/^[0-9]+$/) || value === '') {
            this.setState({ afterErrors: false });
          } else {
            this.setState({ afterErrors: true });
          }
          break;
        case 'interval':
          if (value.match(/^(?=.*[0-9])(?=.*[a-zA-Z])([a-zA-Z0-9]+)$/) || value === '') {
            this.setState({ intervalErrors: false });
          } else {
            this.setState({ intervalErrors: true });
          }
          break;
        default:
          break;
      }
    }
  }
  // Function to validate timestamp and interval inputs in the form

  // Function to handle change as the user inputs
  handleChange(event) {
    const { name, value } = event.target;
    this.setState({ [name]: event.target.value }, () => { this.validateInput(name, value); });
  }
  // Function to handle change as the user inputs


  // Function to validate all inputs and make the API call if the inputs are valid
  handleSubmit(event) {
    const {
      urls,
      before,
      after, interval, beforeErrors, afterErrors, intervalErrors, validInputs, formErrors,
    } = this.state;
    if (urls.length === 0 || before === '' || after === '' || interval === '') {
      this.setState({ formErrors: true });
      if ((urls.length === 1 && urls.shift() === '') || (urls.length === 0)) {
        this.setState({ urlErrors: true });
      } else {
        this.setState({ urlErrors: false });
      }
    } else {
      this.setState({ formErrors: false });
    }
    if (!beforeErrors && !afterErrors && !intervalErrors) {
      this.setState({ validInputs: true });
    } else {
      this.setState({ validInputs: false });
    }
    // Make API call and fetch the response if the inputs are valid
    if (validInputs && !formErrors) {
      this.setState({ submitted: true });
      this.setState({ loading: true });
      event.preventDefault();
      fetch('https://elastictest.herokuapp.com/page_views', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          urls,
          before,
          after,
          interval,
        }),
      }).then((response) => {
        const ret = response.json();
        return ret;
      }).then((response) => {
        this.calculateVals(response);
      });
      // Make API call and fetch the response if the inputs are valid
    } else {
      // Display the error message if the inputs are not valid
      event.preventDefault();
      this.setState({ formErrors: true });
    }
    // Display the error message if the inputs are not valid
  }

  // Function to validate all inputs and make the API call if the inputs are valid
  render() {
    const {
      submitted,
      loading,
      labels,
      values,
      urls,
      formErrors,
      urlErrors,
      beforeErrors,
      afterErrors,
      intervalErrors,
      before,
      after,
      interval,
    } = this.state;

    if (submitted) {
      //  Render loading an  imation if data is still loading but form is submitted
      if (loading) {
        return (
          <div>
            <header className="App-header">
              <h1 className="App-title">Date Histogram Visualisation</h1>
            </header>
            <ReactLoading id="loading" type="balls" color="#000000" height="191px" width="96px" />
          </div>
        );
      }
      // Render graph if data is has loaded

      return (
        <Graph
          labels={labels}
          values={values}
          urls={urls}
        />
      );
    }
    // Render input form if nothing has been submitted
    const errorMessages = [
      'Notice: Please check for errors in the form.',
      'Notice: URL cannot be empty',
      'Notice: Invalid Before Timestamp',
      'Notice: Invalid After Timestamp',
      'Notice: Invalid Interval',
    ];
    const errors = [formErrors, urlErrors, beforeErrors, afterErrors, intervalErrors]
      .map((error, index) => {
        const display = error ? 'block' : 'none';
        return <Error display={display} notice={errorMessages[index]} />;
      });
    return (
      <div className="App">
        <header className="App-header">
          <h1 className="App-title">Date Histogram Visualisation</h1>
        </header>
        <p className="App-intro">
          Please enter the GET request data below:
        </p>
        {errors}
        <form id="bodyForm" onSubmit={this.handleSubmit} onChange={this.handleChange}>
          <label htmlFor="urls">
            URLS:
            <textarea placeholder="Enter URLs separated by commas..." type="text" name="urls" defaultValue={urls} />
          </label>
          <br />
          <label htmlFor="before">
            Before:
            <input placeholder="Enter milliseconds Timestamp..." type="text" name="before" defaultValue={before} />
          </label>
          <br />
          <label htmlFor="after">
            After:
            <input placeholder="Enter milliseconds Timestamp..." type="text" name="after" defaultValue={after} />
          </label>
          <br />
          <label htmlFor="interval">
            Interval:
            <input placeholder="Enter interval such as 2m or 2s..." type="text" name="interval" defaultValue={interval} />
          </label>
          <br />
          <input type="submit" value="Submit" />
        </form>
      </div>
    );
  }
}

export default App;
