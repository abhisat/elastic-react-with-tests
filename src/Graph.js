/**
 * Created by abhisheksatpathy on 17/12/17.
 */
import React, { Component } from 'react';
import './App.css';
import 'plotly.js';

class Graph extends Component {
    constructor(props) {
        super(props);
        this.state = {
            urls: this.props.urls,
            labels: this.props.labels,
            values: this.props.values,
            colours: [],
            loaded: false,
            trace: []
        };
    }

    // Function to draw graph if the DOM element has been mounted
    componentDidMount(){

        // Using Plotly.js for graphs
        var Plotly = require('plotly.js');

        for(var i = 0; i < this.state.labels.length; i++){
            this.state.labels[i] = new Date(this.state.labels[i]).toLocaleTimeString();
        }

        for(var j = 0; j < this.state.labels.length; j++){
            this.state.colours[j] = "rgba(" + Math.random() * (255) + ", " + Math.random() * (255) + "," + Math.random() * (255) + ", 0.7)";
        }

        // Tracing data for each url to be displayed in the graph
        for(var l = 0; l < this.state.urls.length; l++){
            this.state.trace[l] = {
                x: this.state.labels,
                y: this.state.values[l],
                name: this.state.urls[l],
                histnorm: "count",
                marker: {
                    color: this.state.colours[l],
                    line: {
                        color:  "rgba(0, 0, 0, 1)",
                        width: 1
                    }
                },
                opacity: 0.75,
                type: "bar",

            };
        }
        // Layout of the graph
        var layout = {
            barmode: "stack",
            showlegend: true,
            legend: {
                x: 1000,
                y: 1500
            },
            xaxis: {title: "Timestamp"},
            yaxis: {title: "Count"}
        };

        Plotly.newPlot('graph',this.state.trace, layout);

        this.setState({loaded: true});
    }
    // Function to draw graph if the DOM element has been mounted

    render(){
        // Render the graph component once completed drawing
            return(
                <div>
                    <header className="App-header">
                        <h1 className="App-title">Date Histogram Visualisation</h1>
                    </header>
                    <div id="graph"></div>
                </div>
            )
        }
}

export default Graph;
