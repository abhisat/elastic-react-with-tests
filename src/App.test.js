import React from 'react';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import Enzyme, { shallow, mount, sinonwq } from 'enzyme';
import toJson from 'enzyme-to-json';
import canvas from 'canvas';
import jest from 'jest';
import sinon from 'sinon';
import App from './App';
import Error from './Error';
import Graph from './Graph';

// Configure Enzyme Adapter
Enzyme.configure({
  adapter: new EnzymeAdapter(),
});

// Setup App Component proper mount
const setupMountApp = (props = {}, state = {}) => {
  const wrapper = mount(<App {...props} />);
  if (state) wrapper.setState({ ...state });
  return wrapper;
};

// Setup App Component shallow mount
const setupApp = (props = {}, state = {}) => {
  const wrapper = shallow(<App {...props} />);
  if (state) wrapper.setState({ ...state });
  return wrapper;
};

// Setup Graph Component proper mount
const setupMountGraph = (props = {}, state = {}) => {
  const wrapper = mount(<Graph {...props} />);
  if (state) wrapper.setState({ ...state });
  return wrapper;
};

// Setup Graph component shallow mount
const setupGraph = (props = {}, state = {}) => {
  const wrapper = shallow(<Graph {...props} />);
  if (state) wrapper.setState({ ...state });
  return wrapper;
};

// Setup Errors components
const setupErrors = (props = {}) => {
  const wrapper = shallow(<Error {...props} />);
  return wrapper;
};

// Unit Tests to check if the components mount
describe('App components mounts', () => {
  it('App Renders', () => {
    setupApp();
  });

  it('Renders the main div', () => {
    const wrapper = setupMountApp();
    expect(wrapper.find('div').exists()).toBe(true);
    wrapper.unmount();
  });

  it('Graph component mounts', () => {
    const props = {};
    const wrapper = setupMountGraph(props);
    it('Should render the graph with props', () => {
      expect(wrapper.find(Graph).length).toBe(1);
    });
    wrapper.unmount();
  });
});

// Unit tests to check if the individual components render correctly
describe('The components render correctly', () => {
  const wrapper = setupApp();
  describe('The header renders correctly', () => {
    it('Renders the Header text', () => {
      expect(wrapper.find('h1').text()).toBe('Date Histogram Visualisation');
    });

    it('Renders the app introduction', () => {
      expect(wrapper.find('.App-intro').exists()).toBe(true);
    });
  });

  describe('The form renders correctly', () => {
    describe('Renders the form inputs', () => {
      it('Renders the submit button', () => {
        expect(wrapper.find('#submit').exists()).toBe(true);
      });

      it('Renders the URL textarea', () => {
        expect(wrapper.find('#urls').exists()).toBe(true);
      });

      it('Renders the before timestamp textinput', () => {
        expect(wrapper.find('#before').exists()).toBe(true);
      });

      it('Renders the after timestamp textinput', () => {
        expect(wrapper.find('#after').exists()).toBe(true);
      });

      it('Renders the intervals textinput', () => {
        expect(wrapper.find('#interval').exists()).toBe(true);
      });
    });

    describe('Renders the form labels', () => {
      it('Renders the url label', () => {
        expect(wrapper.find('#urlsLabel').exists()).toBe(true);
      });

      it('Renders the before label', () => {
        expect(wrapper.find('#beforeLabel').exists()).toBe(true);
      });

      it('Renders the after label', () => {
        expect(wrapper.find('#afterLabel').exists()).toBe(true);
      });

      it('Renders the interval label', () => {
        expect(wrapper.find('#intervalLabel').exists()).toBe(true);
      });
    });
// The errors are not rendered in the initial render
    describe('Does not render the errors on the first mount', () => {
      it('No error messages rendered', () => {
        expect(wrapper.find('#error').exists()).toBe(false);
      });
    });
  });
});

// Unit tests to check the form validation
describe('Validation Errors display', () => {
  const props = {};
  describe('URL error renders correctly', () => {
    const wrapper = setupApp();
    wrapper.setState({ urlErrors: true }, () => {
      props.key = 'urlErrors';
      props.display = wrapper.state.urlErrors ? 'block' : 'none';
      props.notice = 'Notice: URL cannot be empty';
    });
    wrapper.update();

    it('renders the correct number of errors', () => {
      expect(wrapper.find(Error).length).toEqual(1);
    });
  });

  describe('Before timestamp error renders correctly', () => {
    const wrapper = setupApp();
    wrapper.setState({ beforeErrors: true }, () => {
      props.key = 'beforeErrors';
      props.display = wrapper.state.beforeErrors ? 'block' : 'none';
      props.notice = 'Notice: Invalid Before Timestamp';
    });
    wrapper.update();

    it('renders the correct number of errors', () => {
      expect(wrapper.find(Error).length).toEqual(1);
    });

    it('renders the correct error text', () => {
      expect(wrapper.find(Error).shallow().text()).toBe('Notice: Invalid Before Timestamp');
    });
  });
});

// Tests whether the error component is stateless
describe('Tests if the Error component is stateless', () => {
  const props = {
    key: 0,
    display: 'block',
    notice: 'Notice: URL cannot be empty',
  };
  const wrapper = setupErrors(props);
  const instance = wrapper.instance();

  it('should return null', () => {
    expect(instance).toBe(null);
  });
});

// Unit Tests whether the validateInput functions correctly
describe('Test validateInput function', () => {
  const props = {};
  const state = {
    urls: [],
    before: '',
    after: '',
    interval: '',
    submitted: false,
    loading: false,
    validInputs: false,
    labels: [],
    values: [],
    urlErrors: false,
    beforeErrors: false,
    afterErrors: false,
    intervalErrors: false,
    formErrors: false,
  };
  const wrapper = setupApp(props, state);
  const instance = wrapper.instance();
  describe('validates urls', () => {
    it('tests whether urlErrors state remains unchanged on valid input', () => {
      const name = 'urls';
      const value = 'www.google.com,www.facebook.com,www.yahoo.com,www.bauer-media.com.au';
      expect(wrapper.state('urlErrors')).toBe(false);
      wrapper.setState({ urls: value.split(',') });
      instance.validateInput(name, value);
      expect(wrapper.state('urlErrors')).toBe(false);
    });

    it('tests whether urlsErrors state changes on invalid input', () => {
      const name = 'urls';
      const value = '';
      expect(wrapper.state('urlErrors')).toBe(false);
      wrapper.setState({ urls: value.split(',') });
      instance.validateInput(name, value);
      expect(wrapper.state('urlErrors')).toBe(true);
    });
  });

  describe('validates before timestamp', () => {
    it('tests whether beforeErrors state changes valid input', () => {
      const name = 'before';
      const value = '1554984732';
      expect(wrapper.state('beforeErrors')).toBe(false);
      instance.validateInput(name, value);
      expect(wrapper.state('beforeErrors')).toBe(false);
    });

    it('tests whether beforeErrors state changes on no input', () => {
      const name = 'before';
      const value = '';
      expect(wrapper.state('beforeErrors')).toBe(false);
      instance.validateInput(name, value);
      expect(wrapper.state('beforeErrors')).toBe(false);
    });

    it('tests whether beforeErrors state changes on invalid input', () => {
      const name = 'before';
      const value = 'bvdfjksbvksdbvs';
      expect(wrapper.state('beforeErrors')).toBe(false);
      instance.validateInput(name, value);
      expect(wrapper.state('beforeErrors')).toBe(true);
    });
  });

  describe('validates after timestamp', () => {
    it('tests whether beforeErrors state changes valid input', () => {
      const name = 'after';
      const value = '1554984745';
      expect(wrapper.state('afterErrors')).toBe(false);
      instance.validateInput(name, value);
      expect(wrapper.state('afterErrors')).toBe(false);
    });

    it('tests whether beforeErrors state changes on no input', () => {
      const name = 'after';
      const value = '';
      expect(wrapper.state('afterErrors')).toBe(false);
      instance.validateInput(name, value);
      expect(wrapper.state('afterErrors')).toBe(false);
    });

    it('tests whether beforeErrors state changes on invalid input', () => {
      const name = 'after';
      const value = 'dgcjkdsbvdsv';
      expect(wrapper.state('afterErrors')).toBe(false);
      instance.validateInput(name, value);
      expect(wrapper.state('afterErrors')).toBe(true);
    });
  });

  describe('validates intervals input', () => {
    it('tests whether intervalErrors state changes on valid input', () => {
      const name = 'interval';
      const value = '2s';
      expect(wrapper.state('intervalErrors')).toBe(false);
      instance.validateInput(name, value);
      expect(wrapper.state('intervalErrors')).toBe(false);
    });

    it('tests whether intervalErrors state changes on no input', () => {
      const name = 'interval';
      const value = '';
      expect(wrapper.state('intervalErrors')).toBe(false);
      instance.validateInput(name, value);
      expect(wrapper.state('intervalErrors')).toBe(false);
    });

    it('tests whether intervalErrors state changes on invalid input', () => {
      const name = 'interval';
      const value = 'dvsdvsdvsd';
      expect(wrapper.state('intervalErrors')).toBe(false);
      instance.validateInput(name, value);
      expect(wrapper.state('intervalErrors')).toBe(true);
    });
  });
});

describe('Tests behaviour on submit', () => {
  describe('Submit empty form', () => {
    const props = {};
    const state = {
      urls: [],
      before: '',
      after: '',
      interval: '',
    };
    const wrapper = setupApp(props, state);
    it('should render the error messages for form error and url', () => {
      wrapper.find('form').simulate('submit', {
        preventDefault: () => {
        },
      });
      wrapper.update();
      expect(wrapper.find(Error).length).toEqual(2);
    });
  });

  describe('Submit form with missing inputs', () => {
    const props = {};
    const state = {
      urls: ['www.google.com', 'www.facebook.com'],
      before: '',
      after: '',
      interval: '2s',
    };
    const wrapper = setupApp(props, state);
    it('should render the error messages for form error and url', () => {
      wrapper.find('form').simulate('submit', {
        preventDefault: () => {
        },
      });
      expect(wrapper.find(Error).length).toEqual(1);
    });
  });

  describe('submit form with proper inputs', () => {
    const props = {};
    const state = {
      urls: ['www.google.com', 'www.facebook.com'],
      before: '1551083904',
      after: '1551084478',
      interval: '2s',
    };
    const wrapper = setupMountApp(props, state);
    it('should render the graph component', () => {
      wrapper.find('form').simulate('submit', {
        preventDefault: () => {
        },
      });
      expect(wrapper.find(Graph).length).toEqual(1);
      wrapper.unmount();
    });
  });
});

describe('Test calculateVals function', () => {
  const response = [];
  const props = {};
  const state = {
    urls: [],
    before: '',
    after: '',
    interval: '',
    submitted: false,
    loading: false,
    validInputs: false,
    labels: [],
    values: [],
    urlErrors: false,
    beforeErrors: false,
    afterErrors: false,
    intervalErrors: false,
    formErrors: false,
  };
  const wrapper = setupApp(props, state);
  const instance = wrapper.instance();

  expect(wrapper.state('labels')).toEqual([]);
  expect(wrapper.state('values')).toEqual([]);
  instance.calculateVals(response);
  expect(wrapper.state('labels')).toEqual([]);
  expect(wrapper.state('values')).toEqual([]);
});

// Tests Event Handlers
describe('The event handlers work properly', () => {
  describe('calls handleSubmit', () => {
    const wrapper = setupApp();
    const handleSubmit = sinon.spy(wrapper.instance(), 'handleSubmit');
    wrapper.instance.forceUpdate();
    wrapper.update();
    wrapper.find('form').simulate('submit', {
      preventDefault: () => {
      },
    });
    sinon.assert.calledOnce(handleSubmit);
  });

  describe('calls handleChange', () => {
    const wrapper = setupApp();
    const handleChange = sinon.spy(wrapper.instance(), 'handleChange');
    wrapper.instance.forceUpdate();
    wrapper.update();
    wrapper.find('#urls').simulate('keydown', {
      which: 'a',
      preventDefault: () => {
      },
    });
    sinon.assert.calledOnce(handleChange);
  });
});

// Snapshot tests for comprehensive component comparision
describe('Component snapshots', () => {
  describe('App component snapshot', () => {
    it('should render correctly', () => {
      const wrapper = setupApp();
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('Graph component snapshot', () => {
    it('should render correctly', () => {
      const wrapper = setupGraph();
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('Error components mount', () => {
    it('should render correctly', () => {
      const props = {
        key: 0,
        display: 'block',
        notice: 'Notice: URL cannot be empty',
      };
      const wrapper = setupApp(props);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});
