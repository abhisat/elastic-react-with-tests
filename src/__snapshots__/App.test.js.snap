// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component snapshots App component snapshot should render correctly 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <h1
      className="App-title"
    >
      Date Histogram Visualisation
    </h1>
  </header>
  <p
    className="App-intro"
  >
    Please enter the GET request data below:
  </p>
  <form
    id="bodyForm"
    onChange={[Function]}
    onSubmit={[Function]}
  >
    <label
      htmlFor="urls"
      id="urlsLabel"
    >
      URLS:
      <textarea
        defaultValue={Array []}
        id="urls"
        name="urls"
        placeholder="Enter URLs separated by commas..."
        type="text"
      />
    </label>
    <br />
    <label
      htmlFor="before"
      id="beforeLabel"
    >
      Before:
      <input
        defaultValue=""
        id="before"
        name="before"
        placeholder="Enter milliseconds Timestamp..."
        type="text"
      />
    </label>
    <br />
    <label
      htmlFor="after"
      id="afterLabel"
    >
      After:
      <input
        defaultValue=""
        id="after"
        name="after"
        placeholder="Enter milliseconds Timestamp..."
        type="text"
      />
    </label>
    <br />
    <label
      htmlFor="interval"
      id="intervalLabel"
    >
      Interval:
      <input
        defaultValue=""
        id="interval"
        name="interval"
        placeholder="Enter interval such as 2m or 2s..."
        type="text"
      />
    </label>
    <br />
    <input
      id="submit"
      type="submit"
      value="Submit"
    />
  </form>
</div>
`;

exports[`Component snapshots Error components mount should render correctly 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <h1
      className="App-title"
    >
      Date Histogram Visualisation
    </h1>
  </header>
  <p
    className="App-intro"
  >
    Please enter the GET request data below:
  </p>
  <form
    id="bodyForm"
    onChange={[Function]}
    onSubmit={[Function]}
  >
    <label
      htmlFor="urls"
      id="urlsLabel"
    >
      URLS:
      <textarea
        defaultValue={Array []}
        id="urls"
        name="urls"
        placeholder="Enter URLs separated by commas..."
        type="text"
      />
    </label>
    <br />
    <label
      htmlFor="before"
      id="beforeLabel"
    >
      Before:
      <input
        defaultValue=""
        id="before"
        name="before"
        placeholder="Enter milliseconds Timestamp..."
        type="text"
      />
    </label>
    <br />
    <label
      htmlFor="after"
      id="afterLabel"
    >
      After:
      <input
        defaultValue=""
        id="after"
        name="after"
        placeholder="Enter milliseconds Timestamp..."
        type="text"
      />
    </label>
    <br />
    <label
      htmlFor="interval"
      id="intervalLabel"
    >
      Interval:
      <input
        defaultValue=""
        id="interval"
        name="interval"
        placeholder="Enter interval such as 2m or 2s..."
        type="text"
      />
    </label>
    <br />
    <input
      id="submit"
      type="submit"
      value="Submit"
    />
  </form>
</div>
`;
